### PR TITLE
Block store

### PR DESCRIPTION
**Pull request by tzdybal**
This PR introduces Block store.

Depends on: #40, #41.

### CODE REVIEW

File: p2p/client_test.go [hunk 1]
```diff
@@ -67,7 +67,7 @@ func TestClientStartup(t *testing.T) {
 }
 
 func TestBootstrapping(t *testing.T) {
-	log.SetLogLevel("dht", "INFO")
+	_ = log.SetLogLevel("dht", "INFO")
 	//log.SetDebugLogging()
```
Comment (Wondertan) [inline p2p/client_test.go:71]: I am not sure if we need to keep that logging/debugging code in master actually. I usually remove that code before merging and put it back for local usage only without committing

File: store/blockstore.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
+package store
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/gob"
```
Comment (Wondertan) [inline store/blockstore.go:6]: Bruh, why `gob`?

File: store/blockstore.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+// currently, we're indexing height->hash, and store blocks by hash, but we might as well store by height
+// and index hash->height
+func (bs *DefaultBlockStore) LoadBlock(height uint64) (*types.Block, error) {
+	buf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(buf, height)
+	ikey := append(indexPreffix[:], buf[:]...)
+
+	bs.mtx.RLock()
+	hash, err := bs.db.Get(ikey)
+	bs.mtx.RUnlock()
```
Comment (Wondertan) [inline store/blockstore.go:81]: Why do we need a lock for underlying DB operations if it is thread-safe?

File: store/blockstore.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+func (bs *DefaultBlockStore) Height() uint64 {
+	bs.mtx.RLock()
+	defer bs.mtx.RUnlock()
+	return bs.height
+}
+
+func (bs *DefaultBlockStore) SaveBlock(block *types.Block) error {
+	// TODO(tzdybal): proper serialization & hashing
```
Comment (Wondertan) [inline store/blockstore.go:40]: Why not make it proper already? 
If we building things from scratch here, sticking with easy and performant solutions like [sha256-simd](https://github.com/minio/sha256-simd) from the beginning makes sense.

File: store/blockstore.go (hunk 2 above)
Comment (tzdybal) [reply store/blockstore.go:6]: It works out of the box, without external deps. I know that it will be replaced in the future anyways, so it's the easiest solution.

File: store/blockstore.go (hunk 3 above)
Comment (tzdybal) [reply store/blockstore.go:81]: Good catch. It makes no sense.

File: store/blockstore.go (hunk 4 above)
Comment (tzdybal) [reply store/blockstore.go:40]: Goal of this PR is to introduce block store that works. Using sha256-simd actually makes sense as it's drop-in replacement. By this comment I mostly mean that the serialization and hashing code will definitely be modified in the future.

**Comment by Wondertan** [comment]
@tzdybal, @liamsi, how full blocks are going to be transported between ORU nodes, both aggregators, and simple users and is there a use case for that?

**Comment by liamsi** [comment]
> how full ORU blocks are going to be transported between ORU nodes

That is not specified yet. It will be different from lazyledger-core for sure. If you want to write an ADR for that, that would be more than welcome!

File: store/blockstore.go (hunk 2 above)
Comment (liamsi) [reply store/blockstore.go:6]: I think that is OK, can you open a tracking issue for replacing this with another encoding format (yet to be decided).

File: store/blockstore.go (hunk 3 above)
Comment (liamsi) [reply store/blockstore.go:81]: Can you remove it then?

File: rpcclient/local.go [hunk 5]
```diff
@@ -293,8 +293,33 @@ func (l *Local) TxSearch(ctx context.Context, query string, prove bool, page, pe
 }
 
 func (l *Local) Status(ctx context.Context) (*ctypes.ResultStatus, error) {
-	// needs block store, P2P layer
-	panic("Status - not implemented!")
+	latest, err := l.node.BlockStore.LoadBlock(l.node.BlockStore.Height())
+	if err != nil {
+		// TODO(tzdybal): extract error
```
Comment (liamsi) [inline rpcclient/local.go:298]: What does that mean?

File: store/blockstore.go (hunk 2 above)
Comment (liamsi) [reply store/blockstore.go:6]: tracking issue: https://github.com/lazyledger/optimint/issues/45

File: store/blockstore.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/gob"
+	"sync"
+
+	"github.com/lazyledger/optimint/types"
+	"go.uber.org/multierr"
+	"golang.org/x/crypto/sha3"
```
Comment (liamsi) [inline store/blockstore.go:11]: Can we use sha2 everywhere?
cc @adlerjohn

File: store/blockstore.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+	blockPreffix = [1]byte{1}
+	indexPreffix = [1]byte{2}
+)
+
+type DefaultBlockStore struct {
+	db KVStore
+
+	height uint64
+
+	mtx sync.RWMutex
```
Comment (liamsi) [inline store/blockstore.go:24]: add a comment what this mutex guards

File: store/blockstore.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	"sync"
+
+	"github.com/lazyledger/optimint/types"
+	"go.uber.org/multierr"
+	"golang.org/x/crypto/sha3"
+)
+
+var (
+	blockPreffix = [1]byte{1}
+	indexPreffix = [1]byte{2}
```
Comment (liamsi) [inline store/blockstore.go:16]: Add a comment describing what these prefixes are used for.
Also 
```suggestion
	blockPrefix = [1]byte{1}
	indexPrefix = [1]byte{2}
```

File: store/blockstore.go [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+	if block.Header.Height > bs.height {
+		bs.height = block.Header.Height
+	}
+
+	return err
+}
+
+// TODO(tzdybal): what is more common access pattern? by height or by hash?
+// currently, we're indexing height->hash, and store blocks by hash, but we might as well store by height
+// and index hash->height
```
Comment (liamsi) [inline store/blockstore.go:73]: I think we need both. In optimint you can dictate the access pattern, on the app-side (e.g. via RPC), you can't really know (but you could look into some typical cosmos "apps").

File: store/blockstore.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+	hash, err := bs.db.Get(ikey)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(tzdybal): any better way to convert slice to array?
+	var h [32]byte
+	copy(h[:], hash)
```
Comment (liamsi) [inline store/blockstore.go:87]: I think that is the most common way. You could add a helper method (or methods) if you have to convert back and forth.

File: store/blockstore.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+	return bs.LoadBlockByHash(h)
+}
+
+func (bs *DefaultBlockStore) LoadBlockByHash(hash [32]byte) (*types.Block, error) {
+	key := append(blockPreffix[:], hash[:]...)
+
+	bs.mtx.RLock()
+	blockData, err := bs.db.Get(key)
+	bs.mtx.RUnlock()
```
Comment (liamsi) [inline store/blockstore.go:96]: Does the use of the mutex make sense here?

File: store/blockstore.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+	var block types.Block
+	err = dec.Decode(&block)
+	if err != nil {
+		return nil, err
+	}
+
+	return &block, nil
+}
+
+// TODO(tzdybal): replace with proper hashing mechanism
```
Comment (liamsi) [inline store/blockstore.go:112]: It would be good to decide on all cryptographic primitives in ADR I think.

File: store/blockstore_test.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+	bstore := NewBlockStore()
+	assert.Equal(uint64(0), bstore.Height())
+
+	err := bstore.SaveBlock(&types.Block{
+		Header: types.Header{
+			Height: 1,
+		},
+	})
+	assert.NoError(err)
+	assert.Equal(uint64(1), bstore.Height())
```
Comment (liamsi) [inline store/blockstore_test.go:23]: Better to use table-driven tests and test more cases (instead of only height == 1). Especially edge-cases are interesting too.

File: store/kv.go [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+//
+// KVStore MUST be thread safe.
+type KVStore interface {
+	Get(key []byte) ([]byte, error)     // Get gets the value for a key.
+	Set(key []byte, value []byte) error // Set updates the value for a key.
+	Delete(key []byte) error            // Delete deletes a key.
+}
+
+func NewInMemoryKVStore() KVStore {
+	db, err := badger.Open(badger.DefaultOptions("").WithInMemory(true))
```
Comment (liamsi) [inline store/kv.go:15]: Cool, I didn't know that you can run badger completely in memory, too.

File: rpcclient/local.go (hunk 5 above)
Comment (tzdybal) [reply rpcclient/local.go:298]: It's the reminder to extract/introduce sentinel error value.

File: store/blockstore.go (hunk 11 above)
Comment (tzdybal) [reply store/blockstore.go:96]: Good catch

File: store/blockstore_test.go (hunk 13 above)
Comment (liamsi) [reply store/blockstore_test.go:23]: I think it makes sense to address this in this PR too.

Comment (tzdybal) [reply store/blockstore_test.go:23]: Improved tests are almost ready.

File: store/blockstore_test.go [hunk 15]
```diff
@@ -9,37 +10,115 @@ import (
...
+			getRandomBlock(1, 10),
+			getRandomBlock(2, 20),
+		}},
+		{"blocks out of order", []*types.Block{
+			getRandomBlock(2, 20),
+			getRandomBlock(3, 30),
+			getRandomBlock(4, 100),
+			getRandomBlock(5, 10),
+			getRandomBlock(1, 10),
+		}},
```
Comment (liamsi) [inline store/blockstore_test.go:69]: Much better!

**Review by liamsi — APPROVED** [review]
LGTM
